File: code/llm/python/auth_service.py
"""
Minimal Authentication Microservice in Python
------------------------------------------
Endpoints:
  1) POST /register
  2) POST /login
  3) DELETE /delete

Uses only standard library modules for maximum performance.
Stores users in-memory, returns JWT tokens, and uses HMAC-SHA256 for password hashing.
"""

import json
import hmac
import base64
import hashlib
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse
from typing import Dict, Optional, Tuple
import time

# Configuration
PORT = 3000
SECRET_KEY = b'YOUR_SUPER_SECRET'  # Convert to bytes for HMAC
users: Dict[str, str] = {}  # In-memory user store: {email: hashed_password}

def hash_password(password: str) -> str:
    """Hash a password using HMAC-SHA256."""
    return hmac.new(SECRET_KEY, password.encode(), hashlib.sha256).hexdigest()

def base64url_encode(data: bytes) -> str:
    """Create a base64 URL-safe string."""
    return base64.urlsafe_b64encode(data).rstrip(b'=').decode('utf-8')

def generate_jwt(payload: dict) -> str:
    """Generate a JWT token using HS256 algorithm."""
    header = {'alg': 'HS256', 'typ': 'JWT'}
    
    # Encode header and payload
    header_encoded = base64url_encode(json.dumps(header).encode())
    payload_encoded = base64url_encode(json.dumps(payload).encode())
    
    # Create signature
    signature_input = f"{header_encoded}.{payload_encoded}".encode()
    signature = base64url_encode(
        hmac.new(SECRET_KEY, signature_input, hashlib.sha256).digest()
    )
    
    return f"{header_encoded}.{payload_encoded}.{signature}"

def verify_jwt(token: str) -> Optional[dict]:
    """Verify a JWT token and return the payload if valid."""
    try:
        header_encoded, payload_encoded, signature = token.split('.')
        
        # Verify signature
        signature_input = f"{header_encoded}.{payload_encoded}".encode()
        expected_signature = base64url_encode(
            hmac.new(SECRET_KEY, signature_input, hashlib.sha256).digest()
        )
        
        if signature != expected_signature:
            return None
            
        # Decode payload
        padding = '=' * (-len(payload_encoded) % 4)
        payload_json = base64.urlsafe_b64decode(payload_encoded + padding)
        return json.loads(payload_json)
    except Exception:
        return None

class AuthHandler(BaseHTTPRequestHandler):
    def send_json_response(self, status_code: int, data: dict) -> None:
        """Send a JSON response with given status code and data."""
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())

    def read_json_body(self) -> Tuple[bool, Optional[dict]]:
        """Read and parse JSON body from request."""
        try:
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length)
            return True, json.loads(body)
        except Exception:
            return False, None

    def do_POST(self) -> None:
        """Handle POST requests for /register and /login endpoints."""
        path = urlparse(self.path).path
        
        if path not in ['/register', '/login']:
            self.send_json_response(404, {'error': 'Not found'})
            return
            
        success, data = self.read_json_body()
        if not success or not data or 'email' not in data or 'password' not in data:
            self.send_json_response(400, {'error': 'Invalid JSON or missing fields.'})
            return
            
        email, password = data['email'], data['password']
        
        if path == '/register':
            if email in users:
                self.send_json_response(400, {'error': 'User already exists.'})
                return
                
            users[email] = hash_password(password)
            token = generate_jwt({'email': email})
            self.send_json_response(200, {'token': token})
            
        elif path == '/login':
            stored_hash = users.get(email)
            if not stored_hash or stored_hash != hash_password(password):
                self.send_json_response(401, {'error': 'Invalid credentials.'})
                return
                
            token = generate_jwt({'email': email})
            self.send_json_response(200, {'token': token})

    def do_DELETE(self) -> None:
        """Handle DELETE requests for /delete endpoint."""
        if self.path != '/delete':
            self.send_json_response(404, {'error': 'Not found'})
            return
            
        # Verify Authorization header
        auth_header = self.headers.get('Authorization')
        if not auth_header:
            self.send_json_response(401, {'error': 'Missing Authorization header.'})
            return
            
        parts = auth_header.split()
        if len(parts) != 2 or parts[0] != 'Bearer':
            self.send_json_response(401, {'error': 'Malformed Authorization header.'})
            return
            
        token = parts[1]
        payload = verify_jwt(token)
        if not payload or 'email' not in payload:
            self.send_json_response(401, {'error': 'Invalid or expired token.'})
            return
            
        # Attempt to delete user
        email = payload['email']
        if email in users:
            del users[email]
            self.send_json_response(200, {'success': True})
        else:
            self.send_json_response(400, {'success': False, 'error': 'User not found.'})

def run_server():
    """Start the HTTP server."""
    server = HTTPServer(('localhost', PORT), AuthHandler)
    print(f'Server running on port {PORT}')
    server.serve_forever()

if __name__ == '__main__':
    run_server()

File: code/llm/java/AuthService.java
import com.sun.net.httpserver.HttpServer;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpExchange;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;

import java.util.HashMap;
import java.util.Map;

/**
 * High-Performance Authentication Microservice
 * Uses only native Java APIs for maximum performance:
 * - com.sun.net.httpserver for HTTP server
 * - javax.crypto for cryptography
 * - ConcurrentHashMap for thread-safe in-memory storage
 */
public class AuthService {
    private static final int PORT = 3000;
    private static final String SECRET_KEY = "YOUR_SUPER_SECRET";
    private static final ConcurrentHashMap<String, String> users = new ConcurrentHashMap<>();

    public static void main(String[] args) throws IOException {
        HttpServer server = HttpServer.create(new InetSocketAddress(PORT), 0);
        
        // Create a thread pool with fixed number of threads for better resource management
        server.setExecutor(Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors()));
        
        server.createContext("/register", new RegisterHandler());
        server.createContext("/login", new LoginHandler());
        server.createContext("/delete", new DeleteHandler());
        
        server.start();
        System.out.println("Server started on port " + PORT);
    }

    private static class RegisterHandler implements HttpHandler {
        @Override
        public void handle(HttpExchange exchange) throws IOException {
            if (!"POST".equals(exchange.getRequestMethod())) {
                sendResponse(exchange, 405, "{\"error\": \"Method not allowed\"}");
                return;
            }

            try {
                Map<String, String> requestBody = parseRequestBody(exchange.getRequestBody());
                String email = requestBody.get("email");
                String password = requestBody.get("password");

                if (email == null || password == null) {
                    sendResponse(exchange, 400, "{\"error\": \"Invalid JSON or missing fields\"}");
                    return;
                }

                if (users.containsKey(email)) {
                    sendResponse(exchange, 400, "{\"error\": \"User already exists\"}");
                    return;
                }

                String hashedPassword = hashPassword(password);
                users.put(email, hashedPassword);

                String token = generateJWT(email);
                sendResponse(exchange, 200, "{\"token\": \"" + token + "\"}");
            } catch (Exception e) {
                sendResponse(exchange, 400, "{\"error\": \"Invalid request\"}");
            }
        }
    }

    private static class LoginHandler implements HttpHandler {
        @Override
        public void handle(HttpExchange exchange) throws IOException {
            if (!"POST".equals(exchange.getRequestMethod())) {
                sendResponse(exchange, 405, "{\"error\": \"Method not allowed\"}");
                return;
            }

            try {
                Map<String, String> requestBody = parseRequestBody(exchange.getRequestBody());
                String email = requestBody.get("email");
                String password = requestBody.get("password");

                if (email == null || password == null) {
                    sendResponse(exchange, 400, "{\"error\": \"Invalid JSON or missing fields\"}");
                    return;
                }

                String storedHash = users.get(email);
                if (storedHash == null) {
                    sendResponse(exchange, 401, "{\"error\": \"Invalid credentials\"}");
                    return;
                }

                String givenHash = hashPassword(password);
                if (!storedHash.equals(givenHash)) {
                    sendResponse(exchange, 401, "{\"error\": \"Invalid credentials\"}");
                    return;
                }

                String token = generateJWT(email);
                sendResponse(exchange, 200, "{\"token\": \"" + token + "\"}");
            } catch (Exception e) {
                sendResponse(exchange, 400, "{\"error\": \"Invalid request\"}");
            }
        }
    }

    private static class DeleteHandler implements HttpHandler {
        @Override
        public void handle(HttpExchange exchange) throws IOException {
            if (!"DELETE".equals(exchange.getRequestMethod())) {
                sendResponse(exchange, 405, "{\"error\": \"Method not allowed\"}");
                return;
            }

            String authHeader = exchange.getRequestHeaders().getFirst("Authorization");
            if (authHeader == null) {
                sendResponse(exchange, 401, "{\"error\": \"Missing Authorization header\"}");
                return;
            }

            String[] parts = authHeader.split(" ");
            if (parts.length != 2 || !"Bearer".equals(parts[0])) {
                sendResponse(exchange, 401, "{\"error\": \"Malformed Authorization header\"}");
                return;
            }

            try {
                String email = verifyJWT(parts[1]);
                if (email == null) {
                    sendResponse(exchange, 401, "{\"error\": \"Invalid or expired token\"}");
                    return;
                }

                if (users.remove(email) != null) {
                    sendResponse(exchange, 200, "{\"success\": true}");
                } else {
                    sendResponse(exchange, 400, "{\"success\": false, \"error\": \"User not found\"}");
                }
            } catch (Exception e) {
                sendResponse(exchange, 401, "{\"error\": \"Invalid token\"}");
            }
        }
    }

    private static Map<String, String> parseRequestBody(InputStream is) throws IOException {
        byte[] buffer = new byte[1024];
        int bytesRead;
        StringBuilder requestBody = new StringBuilder();
        while ((bytesRead = is.read(buffer)) != -1) {
            requestBody.append(new String(buffer, 0, bytesRead));
        }

        // Simple JSON parsing for {"email": "value", "password": "value"}
        String body = requestBody.toString();
        Map<String, String> result = new HashMap<>();
        
        // Remove brackets and split by comma
        body = body.replaceAll("[{}\"]", "");
        String[] pairs = body.split(",");
        
        for (String pair : pairs) {
            String[] keyValue = pair.split(":");
            if (keyValue.length == 2) {
                result.put(keyValue[0].trim(), keyValue[1].trim());
            }
        }
        
        return result;
    }

    private static void sendResponse(HttpExchange exchange, int statusCode, String response) throws IOException {
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        byte[] responseBytes = response.getBytes(StandardCharsets.UTF_8);
        exchange.sendResponseHeaders(statusCode, responseBytes.length);
        try (OutputStream os = exchange.getResponseBody()) {
            os.write(responseBytes);
        }
    }

    private static String hashPassword(String password) {
        try {
            Mac hmac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKey = new SecretKeySpec(SECRET_KEY.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            hmac.init(secretKey);
            byte[] hash = hmac.doFinal(password.getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(hash);
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            throw new RuntimeException("Error hashing password", e);
        }
    }

    private static String generateJWT(String email) {
        try {
            // Create JWT header and payload
            String header = Base64.getUrlEncoder().withoutPadding().encodeToString(
                "{\"alg\":\"HS256\",\"typ\":\"JWT\"}".getBytes(StandardCharsets.UTF_8));
            String payload = Base64.getUrlEncoder().withoutPadding().encodeToString(
                String.format("{\"email\":\"%s\"}", email).getBytes(StandardCharsets.UTF_8));

            // Create signature
            String signatureInput = header + "." + payload;
            Mac hmac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKey = new SecretKeySpec(SECRET_KEY.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            hmac.init(secretKey);
            byte[] signature = hmac.doFinal(signatureInput.getBytes(StandardCharsets.UTF_8));
            String encodedSignature = Base64.getUrlEncoder().withoutPadding().encodeToString(signature);

            return header + "." + payload + "." + encodedSignature;
        } catch (Exception e) {
            throw new RuntimeException("Error generating JWT", e);
        }
    }

    private static String verifyJWT(String token) {
        try {
            String[] parts = token.split("\\.");
            if (parts.length != 3) return null;

            // Verify signature
            Mac hmac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKey = new SecretKeySpec(SECRET_KEY.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            hmac.init(secretKey);
            byte[] signatureBytes = hmac.doFinal((parts[0] + "." + parts[1]).getBytes(StandardCharsets.UTF_8));
            String computedSignature = Base64.getUrlEncoder().withoutPadding().encodeToString(signatureBytes);

            if (!computedSignature.equals(parts[2])) return null;

            // Decode payload
            String payload = new String(Base64.getUrlDecoder().decode(parts[1]), StandardCharsets.UTF_8);
            // Simple JSON parsing to extract email
            return payload.split("\"email\":\"")[1].split("\"")[0];
        } catch (Exception e) {
            return null;
        }
    }
}
